File: client/src/types/listing.ts
import type { PropertyLocation } from "./location"
import type { Reservation } from "./reservation"
export type ListingUserFavorites = {
  userId: number
}
interface User {
  name: string
  email: string
}
export interface Listing {
  id: number | string
  title: string
  description: string
  imageSrc: string
  country: string
  category: string
  price: number
  location: PropertyLocation
  favorites: ListingUserFavorites[]
  guestCount: number
  roomCount: number
  bathroomCount: number
  reservation: Reservation[]
  user: User
}

File: server/internal/db/listing.go
package db

import (
	"gorm.io/gorm"
)

type Listing struct {
	gorm.Model
	Title         string
	Description   string
	Imagesrc      string
	Category      string
	Roomcount     int
	BathroomCount int
	GuestCount    int
	Location      string
	Price         float32
	Country       string
	UserId        uint
	User          User
	Reservations  []Reservation
	Favorites     []ListingFavorite
}

type ListingFavorite struct {
	gorm.Model `json:"-"`
	ListingID  uint `json:"-"`
	Listing    Listing
	UserId     uint `json:"userId"`
}

func (l *Listing) Favorite(user uint) {
	Db.Create(&ListingFavorite{ListingID: l.ID, UserId: user})
}
